Firestore/CMakeLists.txt

```diff
@@ -15,6 +15,53 @@
 cmake_minimum_required(VERSION 2.8.11)
 project(firestore C CXX)
 
+option(ASAN "Build with Address sanitizer (mutually exculsive with TSAN)" OFF)
+# TODO(varconst): msan
+# Memory sanitizer is more complicated:
+# - it requires all dependencies to be compiled with msan enabled (see
+#   https://github.com/google/sanitizers/wiki/MemorySanitizerLibcxxHowTo);
+# - AppleClang doesn't support it.
+option(TSAN "Build with Thread sanitizer (mutually exculsive with ASAN)" OFF)
+option(UBSAN "Build with Undefined behavior sanitizer" OFF)
+
+if (CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
+  set(HAVE_SANITIZERS 1)
+endif()
+if(ASAN OR TSAN OR UBSAN)
+  set(USE_SANITIZERS 1)
+endif()
+
+if (NOT HAVE_SANITIZERS AND USE_SANITIZERS)
+  message(FATAL_ERROR "Only Clang and GCC support sanitizers.")
+endif()
+
+function(add_to_compile_and_link_flags flag)
+  set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} ${flag}" PARENT_SCOPE)
+  set(CMAKE_LINKER_FLAGS "${CMAKE_LINKER_FLAGS} ${flag}" PARENT_SCOPE)
+endfunction()
+
+if (HAVE_SANITIZERS AND USE_SANITIZERS)
```

Comment:
This logic seems more complicated than it needs to be. It seems like you could just have

```
if(CMAKE_CXX_COMPILER_ID MATCHES "Clang" OR CMAKE_CXX_COMPILER_ID STREQUAL "GNU")
  if(ASAN)
    // add flags
  endif()

  if(TSAN)
    if(ASAN OR UBSAN)
      message(FATAL_ERROR "Cannot combine tsan with other santizers")
    endif()
    // add flags
  endif()

  if(UBSAN)
    // add flags
  endif()
else()
  if(ASAN OR TSAN OR UBSAN)
    message(FATAL_ERROR "Only Clang and GCC support sanitizers")
  endif()
endif()
```

This eliminates `HAVE_SANITIZERS` and `USE_SANITIZERS`  and all these extra blocks that check conditions you're effectively duplicating in here through nesting.

Reply:
Done. The intention behind those named variables was to make things clearer, but I think the shorter version is at least as readable.